compiler/rustc_middle/src/mir/syntax.rs

```diff
@@ -1196,17 +1196,26 @@ pub enum AggregateKind<'tcx> {
     Array(Ty<'tcx>),
     Tuple,
 
-    /// The second field is the variant index. It's equal to 0 for struct
-    /// and union expressions. The fourth field is
-    /// active field number and is present only for union expressions
-    /// -- e.g., for a union expression `SomeUnion { c: .. }`, the
-    /// active field index would identity the field `c`
-    Adt(DefId, VariantIdx, SubstsRef<'tcx>, Option<UserTypeAnnotationIndex>, Option<usize>),
-
-    // Note: We can use LocalDefId since closures and generators a deaggregated
-    // before codegen.
-    Closure(LocalDefId, SubstsRef<'tcx>),
-    Generator(LocalDefId, SubstsRef<'tcx>, hir::Movability),
+    Struct {
+        def_id: DefId,
+        substs: SubstsRef<'tcx>,
+        user_ty: Option<UserTypeAnnotationIndex>,
+    },
+    Enum {
+        def_id: DefId,
+        substs: SubstsRef<'tcx>,
+        variant_index: VariantIdx,
+        user_ty: Option<UserTypeAnnotationIndex>,
+    },
+    Union {
+        def_id: DefId,
+        substs: SubstsRef<'tcx>,
+        active_field_index: Field,
+        user_ty: Option<UserTypeAnnotationIndex>,
+    },
+
+    Closure(DefId, SubstsRef<'tcx>),
+    Generator(DefId, SubstsRef<'tcx>, hir::Movability),
```

Comment:
Please make this split a separate PR, it seems like something we should discuss and review on its own

Reply:
Is this split even a good idea? At least in Miri, I think some of the code duplication is caused by the redundancy in the Struct/Enum/Union cases. Generally the ADT handling in rustc is treating structs as "1-variant enums" and that tends to avoid code duplication. Why diverge from that pattern here?

Reply:
I don't know the motivation for it, but whatever it is, it seems orthogonal to changing deaggregation.

Reply:
We would appreciate this change in Prusti, but I don't know how useful/annoying this is for other consumers of the MIR.

Reply:
I did this split in a separate commit to have this discussion.

The three cases are deaggegated/interpreted/codegened in very different ways, so keeping them merged required to disentangle them manually by checking `adt_def(def_id).is_enum()`.

For instance, deaggregation considered the struct and union field accesses to be direct, without the implicit downcast to variant 0.

If we stop deaggregating, this split is probably not that useful: the interpreter and codegen could perform the implicit downcast to variant 0 in all cases, and it should be no-op for struct and enums.

@Aurel300 could you elaborate on the patterns that this enables/inhibits?

Reply:
In our encoding we also treat struct and enums differently (and we don't yet support unions), so it's common to see this in one `match` arm:

https://github.com/viperproject/prusti-dev/blob/17b071c41c1b502d739ccf414a2445a1bad8a3f6/prusti-viper/src/encoder/snapshot/encoder.rs#L715

And in another:

https://github.com/viperproject/prusti-dev/blob/17b071c41c1b502d739ccf414a2445a1bad8a3f6/prusti-viper/src/encoder/snapshot/encoder.rs#L744

Reply:
@Aurel300 your examples show usage of `TyKind::Adt`. This discussion is about `mir::AggregateKind::Adt`. Am I missing something?

Reply:
> The three cases are deaggegated/interpreted/codegened in very different ways, 

Are they? At least for struct and enum, we iterate over all the fields of one variant of an ADT, and assign them. That's pretty similar.

Reply:
@cjgillot Ah, disregard my comment: I incorrectly assumed the change was on `TyKind`. For encoding a `AggregateKind` we still need some special logic *only if* the ADT is an enum, but then perform common actions in both enum and struct cases.